### ISSUE TITLE
Connection Timeouts in SmartglassClient

### ISSUE THREAD

**Issue by chefbennyj1** [OpenXbox/xbox-smartglass-csharp#50]
Interesting. The first project I built using the SmartglassClient seemed to connect to my Xbox console easily. I was able to register ACK messages in the console readout.
The second project I built seems to be having connection issues, and the first project doesnt get readouts anymore.

The client object will always error with either a null reference exception a timeout exception.

However, using the ConnectAsync and immediately calling the clients void function to turn off the console does work. 

I'm wondering if it has to do with the certificate and authentication with the Xbox console?

**Comment by chefbennyj1** [comment]
There maybe some issue with the async connect method.

The client will not timeout if you call Result with await operator.

This is an incorrect way for calling an asynchronous method.

**Comment by tuxuser** [comment]
What you are saying is:

https://github.com/OpenXbox/xbox-smartglass-csharp/blob/master/SmartGlass/SmartGlassClient.cs#L64
```cs
var response = await TaskExtensions.WithRetries(() =>
    transport.WaitForMessageAsync<ConnectResponseMessage>(
        connectTimeout,
        () => connectFunc().GetAwaiter().GetResult()),
    connectRetries);
```

should be

```cs
var response = await TaskExtensions.WithRetries(() =>
    transport.WaitForMessageAsync<ConnectResponseMessage>(
        connectTimeout,
        async () => await connectFunc()),
    connectRetries);
```

Is that correct?

**Comment by joelday** [comment]
async/await, tasks, etc. have some quirky (or at least difficult to remember) behavior under the hood: https://msdn.microsoft.com/en-us/magazine/jj991977.aspx. I probably did a handful of things that can behave strangely due to some of these nuances.

Short story is, a Task<T> you get from an async method that awaits, the one you await on, and one you get from Task.Run can do different things.

Here's an example of where awaiting on the return Task instead of waiting on a new Task screwed me: https://github.com/joelday/papyrus-lang/blob/master/src/DarkId.Papyrus.Server/ProjectProgramOptionsProvider.cs#L40

**Comment by joelday** [comment]
@chefbennyj1 Thanks for catching this. TBH I'd like to audit the entire codebase for dangerous asynchronous code.

**Comment by joelday** [comment]
@chefbennyj1 Do you mind testing this PR to see if the issue still reproduces for you?

**Comment by joelday** [comment]
Should be fixed by #59
